File: backend/internal/models/franchise.go
package models

import (
	"time"

	"github.com/lib/pq"
)

type Franchise struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	CreatedAt   time.Time      `json:"createdAt"`
	UpdatedAt   time.Time      `json:"updatedAt"`
	Name        string         `gorm:"uniqueIndex;not null" json:"name"`
	NameI18n    *LocalizedString `gorm:"type:jsonb" json:"nameI18n,omitempty"`
	Description *string        `json:"description"`
	Poster      *string        `json:"poster"`
	Aliases     pq.StringArray `gorm:"type:text[]" json:"aliases"`
	Links       []FranchiseMediaLink `gorm:"foreignKey:FranchiseID" json:"links,omitempty"`
}

type FranchiseMediaLink struct {
	ID            uint              `gorm:"primaryKey" json:"id"`
	CreatedAt     time.Time         `json:"createdAt"`
	UpdatedAt     time.Time         `json:"updatedAt"`
	FranchiseID   uint              `gorm:"not null" json:"franchiseId"`
	Franchise     Franchise         `gorm:"foreignKey:FranchiseID" json:"franchise,omitempty"`
	FromMediaType MediaType         `gorm:"not null" json:"fromMediaType"`
	FromMediaID   uint              `gorm:"not null" json:"fromMediaId"`
	ToMediaType   MediaType         `gorm:"not null" json:"toMediaType"`
	ToMediaID     uint              `gorm:"not null" json:"toMediaId"`
	RelationType  MediaRelationType `gorm:"not null" json:"relationType"`
	OrderNumber   *int              `json:"orderNumber"`
	IsCanon       *bool             `json:"isCanon"`
	Note          *string           `json:"note"`
}

File: frontend/src/api/franchise.ts
import apiClient from '@/api/client'
import type { LocalizedString, PaginatedResponse } from '@/types'

export interface FranchiseLinkItem {
  id: number
  franchiseId: number
  franchiseName: string
  franchiseNameI18n?: LocalizedString
  relationType: string
  relatedType: string
  relatedMediaId: number
  relatedTitle: string
  relatedTitleI18n?: LocalizedString
  relatedPoster?: string | null
  relatedRating?: number | null
  listStatus?: string | null
}

export interface Franchise {
  id: number
  name: string
  nameI18n?: LocalizedString
  description?: string
  poster?: string
  aliases?: string[]
  links?: FranchiseMediaLink[]
}

export interface FranchiseMediaLink {
  id: number
  franchiseId: number
  fromMediaType: string
  fromMediaId: number
  toMediaType: string
  toMediaId: number
  relationType: string
  orderNumber?: number
  note?: string
}

export function getFranchiseLinksByMedia(mediaType: string, mediaId: number): Promise<FranchiseLinkItem[]> {
  return apiClient
    .get<FranchiseLinkItem[]>('/franchises/by-media', { params: { mediaType, mediaId } })
    .then((r) => r.data)
}

export const franchiseApi = {
  getList: (page = 1, pageSize = 50) =>
    apiClient.get<PaginatedResponse<Franchise>>('/franchises', { params: { page, pageSize } }).then((r) => r.data),
  search: (q: string) =>
    apiClient.get<Franchise[]>('/franchises/search', { params: { q: q.trim() } }).then((r) => r.data),
  getOne: (id: number) => apiClient.get<Franchise>(`/franchises/${id}`).then((r) => r.data),
  getMedia: (id: number) => apiClient.get<FranchiseMediaLink[]>(`/franchises/${id}/media`).then((r) => r.data),
  create: (data: {
    name: string
    nameI18n?: LocalizedString
    description?: string
    poster?: string
    aliases?: string[]
  }) => apiClient.post<Franchise>('/admin/franchises', data).then((r) => r.data),
  update: (
    id: number,
    data: { name?: string; nameI18n?: LocalizedString; description?: string; poster?: string; aliases?: string[] },
  ) => apiClient.put<Franchise>(`/admin/franchises/${id}`, data).then((r) => r.data),
  delete: (id: number) => apiClient.delete(`/admin/franchises/${id}`),
  addLink: (
    franchiseId: number,
    data: {
      fromMediaType: string
      fromMediaId: number
      toMediaType: string
      toMediaId: number
      relationType: string
      orderNumber?: number
      note?: string
    },
  ) => apiClient.post<FranchiseMediaLink>(`/admin/franchises/${franchiseId}/links`, data).then((r) => r.data),
  updateLink: (linkId: number, data: { relationType?: string; orderNumber?: number; note?: string }) =>
    apiClient.put<FranchiseMediaLink>(`/admin/franchises/links/${linkId}`, data).then((r) => r.data),
  deleteLink: (linkId: number) => apiClient.delete(`/admin/franchises/links/${linkId}`),
}
